### ISSUE TITLE
CS0219 not working for non-primitive types

### ISSUE THREAD

**Issue by areevesat** [dotnet/roslyn#81743]
**Version Used**: 
.NET 9.0.308

**Steps to Reproduce**:

1. Activate analysis rule CS0219, set to warning.
2. Create code with assignment to a non-primitive variable that is not used later.

In the snippet and screenshot below, the initializations of num1 and num2 are correctly flagged by CS0219.  The initialization of violation2 should also be flagged, but it is not.  The assignment to violation3 is flagged by IDE0059 (which should also catch violation2, so that one isn't working properly either.)

```C#
public async Task<Violation?> TestMethod(Guid violationId, CancellationToken cancellationToken)
{
    var violation1 = await this.violationStore.GetById(violationId, cancellationToken);
    int num1 = 0;

    try
    {
        var violation2 = await this.violationStore.GetById(violationId, cancellationToken);
        int num2 = 1;
    }
    catch (Exception)
    {
        return null;
    }

    var violation3 = await this.violationStore.GetById(violationId, cancellationToken);

    return violation1;
}
```

<img width="804" height="354" alt="Image" src="https://github.com/user-attachments/assets/096fe50f-6b7d-48c3-9aee-802ff4625499" />

**Diagnostic Id**:
CS0219: The variable 'variable' is assigned but its value is never used

**Expected Behavior**: Visual Studio should warn that the assigned value is unused, even for non-primitive types.

**Actual Behavior**: No warning is issued for the primitive types.

**Comment by areevesat** [comment]
Created at the same time as #81742.

**Comment by areevesat** [comment]
This appears to be the known functionality, so this may be a feature request rather than a bug.  There is currently no reliable way to issue warnings for all unused variables, and this is one of the many rules that _should_ perform this behavior but doesn't quite.

**Comment by jcouv** [comment]
~~Closing as duplicate of https://github.com/dotnet/roslyn/issues/65220~~

**Comment by areevesat** [comment]
@jcouv This is not a duplicate of #65220, this one is referring to a different rule, CS0219.

**Comment by areevesat** [comment]
Modifying this functionality in code seems like it would be a fairly simple matter of removing the special check for reference types in WriteConsideredUse within Compilers\CSharp\Portable\FlowAnalysis\DefiniteAssignment.cs (I have not checked for VB).

```C#
internal static bool WriteConsideredUse(TypeSymbol type, BoundExpression value)
{
    [...]
    if ((object)type != null && type.IsReferenceType &&
        type.SpecialType != SpecialType.System_String &&
        type is not ArrayTypeSymbol { IsSZArray: true, ElementType.SpecialType: SpecialType.System_Byte })
    {
        return value.ConstantValueOpt != ConstantValue.Null;
    }
    [...]
}
```

However, comments within that file note that assignment to some variables is considered as having read the variable, motivated primarily by giving the ability to see the value of the variable in a debugger.

In my opinion, allowing certain debugging conveniences should not override the need for an analyzer to flag unused variables, which often indicate a logic error.  In my experience, if I am debugging and need to do some bad code practices to get a value, I don't mind seeing a warning on my bad code, and indeed I prefer to have a warning to remind me that I need to come back and remove it.  Or if in a warning-as-error environment, I could always #pragma warning disable CS0219 on my debugging code.

The stated motivations in a comment (from 2014):

>The debugger does not make it easy to see the returned value of a method. Often a call whose returned value would normally be discarded is written into a local variable so that it can be easily inspected in the debugger.

I believe Visual Studio now does show a "returning" value when returning from a method.  And as stated above, it seems fine to throw a warning if someone is temporarily saving a value for debugging purposes, which they can either pragma disable or keep the warning and remove the code later.

> An otherwise unread local variable that contains a reference to an object can keep the object alive longer, particularly if the jitter is not optimizing the lifetimes of locals. (Because, for example, the debugger is running.) Again, this can be useful when debugging because an otherwise unused object might be finalized later, allowing the developer to more easily examine its state.

This use case I am less familiar with, but I again think #pragma warning disable would make sense if this is needed.

> A developer who wishes to deliberately discard a value returned by  a method can do so in a self-documenting manner via "var unread = M();"

Since this comment was written, C# has added the underscore discard functionality.  So this motivation is no longer applicable.

Ultimately, I think most developers using an analyzer would want unused reference type variables to be flagged, and specialized usage for debugging purposes could be suppressed as needed.
